### PR TITLE
Status led pattern: disabled arm reason

### PR DESCRIPTION
**Pull request by gaelj**
Use Warning LED flash count to inform arming prevention reason.

| Reason for disabled Arming               | LED Flashes |
|------------------------------------------|-------------|
| CLI is active in the configurator        | 2           |
| Failsafe mode is active                  | 3           |
| The aircraft has landed in failsafe mode | 3           |
| Maximum arming angle is exceeded         | 4           |
| Calibration is active                    | 5           |
| The system is overloaded                 | 6           |

See PR  #2182
Implements @ledvinap's travis fix.

Looking forward to read your comments.

### CODE REVIEW

**Comment by hydra** [comment]
can you move the travis/build related changes to a different PR (if needed) and rebase against current master.  hopefully the changes from @ledvinap that i just merged should fix the build issues.

I will have a look at this, i think it would be very helpful to the users.

soooo much to do!

**Comment by gaelj** [comment]
Hello @hydra. The travis build changes I incorporated were only there to make it pass the unit tests at the time of posting.
No need for a separate PR since the fix in now merged.

**Comment by gaelj** [comment]
Rebased.

File: src/main/io/beeper.h [hunk 1]
```diff
@@ -47,3 +47,4 @@ uint32_t getArmingBeepTimeMicros(void);
 beeperMode_e beeperModeForTableIndex(int idx);
 const char *beeperNameForTableIndex(int idx);
 int beeperTableEntryCount(void);
+uint8_t beeperIsOn;
```
Comment (ledvinap) [inline src/main/io/beeper.h]: `extern` here ...

File: src/main/io/statusindicator.c [hunk 2]
```diff
@@ -32,59 +55,76 @@ static uint32_t warningLedTimer = 0;
...
+    if (isCalibrating()) {
+        return WARNING_ARM_PREV_CALIB;
+    }
+    if (rcModeIsActive(BOXFAILSAFE) || failsafePhase() == FAILSAFE_LANDED) {
+        return WARNING_ARM_PREV_FAILSAFE;
+    }
+    if (!imuIsAircraftArmable(armingConfig()->max_arm_angle)) {
+        return WARNING_ARM_PREV_ANGLE;
+    }
+    if (cliMode == 1) {
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: if is more robust (and common) to test for non-zero `if(cliMode)`

File: src/main/io/statusindicator.c [hunk 3]
```diff
@@ -32,59 +55,76 @@ static uint32_t warningLedTimer = 0;
...
-    WARNING_LED_FLASH
+    WARNING_ARM_PREV_CLI,       // 2 flashes - CLI active in the configurator
+    WARNING_ARM_PREV_FAILSAFE,  // 3 flashes - Failsafe mode
+    WARNING_ARM_PREV_ANGLE,     // 4 flashes - Maximum arming angle exceeded
+    WARNING_ARM_PREV_CALIB,     // 5 flashes - Calibration active
+    WARNING_ARM_PREV_OVERLOAD   // 6 flashes - System overload
 } warningLedState_e;
 
 static warningLedState_e warningLedState = WARNING_LED_OFF;
+static uint8_t flashsLeft = 0;
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: 'flashes'

**Comment by ledvinap** [comment]
@gaelj : Sorry, I don't like the code. You are trying to hard to keep old code, making things more complicated.

With `warningLedFlash` removed, the code got a lot simpler.
Just implement
`warningLedBeeper(bool on);` that will set warning led it if is not used (for warning blinks).
Blink code can set correct state when starting blink, so it's ok to ignore start/end of beeper.

It's probably better to sample warning state only when blinking is done.

It may make sense to use `uint32_t` mask for blinks - use last bit as marker (not blinking it), then on each tick shift one bit to led. It will allow more complicated blink combinations (Morse anyone ;-) ) and all blink code will be in one switch .. Testing for active blinking is trivial, all state is in single variable ..


Another option is to merge warning led into beeper code - use existing code, just mark if output is beeper/led or both. But beeper code will probably need some cleanup too ...

**Comment by gaelj** [comment]
@ledvinap thanks for your review.
I've implemented your remarks, expect for the `uint32_t` mask, as I think it would increase RAM footprint (each blinking mode would have it's own 32 bits, right ?) as well as increase interpretation complexity for users.
If I'm wrong, could you explain a little further what you have in mind ?

Merging into beeper code was indeed a possibility, but @hydra's request was to show it as warning LED blink patterns.

File: src/main/io/statusindicator.c [hunk 4]
```diff
@@ -20,71 +20,106 @@
...
+    }
+    return ARM_PREV_NONE;
 }
 
 void warningLedRefresh(void)
 {
-    switch (warningLedState) {
-        case WARNING_LED_OFF:
+    if (flashesLeft == 0) {
+        armingPreventionReason = getArmingPreventionReason();
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: you can set `flashesLeft` (persistent state) here. It will simplify code structure.

(yours is:
```
if(A)
   B = state();
if(B)
  if(A)
     A = fn(B);
```
)

File: src/main/io/statusindicator.c [hunk 5]
```diff
@@ -20,71 +20,106 @@
...
+    }
+
+    if (armingPreventionReason > ARM_PREV_NONE) {
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: `flashesLeft` may be uses as state, no need to keep armingPreventionReason too.

File: src/main/io/statusindicator.c [hunk 6]
```diff
@@ -20,71 +20,106 @@
...
+            flashesLeft = 2 * (armingPreventionReason + 1);
             LED0_OFF;
-            break;
-        case WARNING_LED_ON:
-            LED0_ON;
-            break;
-        case WARNING_LED_FLASH:
+        } else {
+            flashesLeft--;
             LED0_TOGGLE;
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: This toggle uses LED as hidden state, which can lead to problems. 
```
if(flashesLeft) {
  if((flashesLeft % 2) == 0)
    LED0_ON;
  else
    LED0_OFF;
  flashesLeft--;
}
```

**Comment by ledvinap** [comment]
The bitmask could be implemented as:
```
static uint32_t blinkMask = 0;

static const uint32 blinkTab[] = {
[ARM_PREV_NONE] = 0b0,
[ARM_PREV_CLI] = 0b1000101;  // *.*...
[ARM_PREV_FAILSAFE] = 0b100010101;  // *.*.*...
// it is possible to use long blinks
[ARM_WARN_X == 0x10001011   // long-short-delay

void warningLedRefresh(void)
{
    if(!blinkMask) {
      blinkMask = blinkTab[getArmingPreventionReason()];
    }
    if(blink_mask) {
      if(blinkMask == 1) {   // special-case for end-of-sequence marker
        LED0_OFF;
      } else {
         if(blinkMask & 1) 
           LED0_ON;
         else
            LED0_OFF;   
      }
      blinkMask >>= 1;
    }
}

**Comment by gaelj** [comment]
Nice. Now I understand the part about that LSB marker in the 32 bit word.
I saved a few bytes by simply using :
```
typedef enum {
    ARM_PREV_NONE       = 0,
    ARM_PREV_CLI        = 0b0000000010101,  // 2 flashes - CLI active in the configurator
    ARM_PREV_FAILSAFE   = 0b0000001010101,  // 3 flashes - Failsafe mode
    ARM_PREV_ANGLE      = 0b0000101010101,  // 4 flashes - Maximum arming angle exceeded
    ARM_PREV_CALIB      = 0b0010101010101,  // 5 flashes - Calibration active
    ARM_PREV_OVERLOAD   = 0b1010101010101   // 6 flashes - System overload
} armingPreventedReason_e;
```

**Comment by ledvinap** [comment]
@gaejl: With bitmask you can encode pause between blinks into blink code, so there is no need for long/short delay: <1-terminator><000-delay><101-pattern>

**Comment by gaelj** [comment]
This should now be a winner. @ledvinap if you are happy with the way it now is I'll rebase & squash.

File: src/main/io/statusindicator.c [hunk 7]
```diff
@@ -20,71 +20,103 @@
...
+            LED0_ON;
+        else
+            LED0_OFF;
+        blinkMask >>= 1;
+    }
 
-void warningLedFlash(void)
-{
-    warningLedState = WARNING_LED_FLASH;
+    uint32_t now = micros();
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: This may be moved out of this function (and into scheduler eventually)

**Comment by ledvinap** [comment]
Great work. I hope I did not nag you too much ..

**Comment by gaelj** [comment]
Absolutely not ! Thank you again for your time @ledvinap.
@hydra we have a merge candidate as soon as you feel like it.

**Comment by gaelj** [comment]
@hydra, now it's ready.

File: src/main/io/statusindicator.c [hunk 8]
```diff
@@ -20,71 +20,101 @@
...
 
 void warningLedUpdate(void)
 {
     uint32_t now = micros();
-
-    if ((int32_t)(now - warningLedTimer) < 0) {
-        return;
+    if ((int32_t)(now - warningLedTimer) > 0) {
+        warningLedRefresh();
+        warningLedTimer = now + WARNING_LED_BLINK_DELAY;
```
Comment (ledvinap) [inline src/main/io/statusindicator.c]: One thing, only for reference:
other possibility is to use `warningLedTimer += WARNING_LED_BLINK_DELAY;`. 
That way timing is more accurate (difference between `now` and `warningLedTimer` is not discarded as in current code. It is even possible to use fractional format for time and invocation intervals get rounded correctly.
Big disadvantage is the you must sanitize `warningLedTimer` (called too late/invalid value).

For blinking, timing from last invocation (current code) is usually better (blink lengths are consistent, not absolute time).

**Comment by hydra** [comment]
ok, this will be very useful for many, i will merge this ahead of the OSD code which needs something similar.  I will update the OSD code to use this now.

**Comment by hydra** [comment]
although i just merged this, i don't like the tight coupling between arm prevention reason and blink masks here:

```
blinkMask = getArmingPreventionReason();
```

IMHO blink masks is an implementation detail of the blinker code, arming prevention reason is an output of system state.

**Comment by hydra** [comment]
@ledvinap and @gaelj 

although i just merged this, i don't like the tight coupling between arm prevention reason and blink masks here:

```
blinkMask = getArmingPreventionReason();
```

IMHO blink masks is an implementation detail of the blinker code, arming prevention reason is an output of system state.

**Comment by hydra** [comment]
i think to fix it, just store the blink masks in an array and index into it based on the reason.   however that would result in fragility if the reason list changed.

however, that could and should be mitigated by a unit test that asserts the array of blink masks is the same as the amount of reason codes.

this is where I would like native language support for maps in C would be useful...

**Comment by hydra** [comment]
lastly, it feels like there is a bit of duplication between the arming code and the getArmingPreventionReason().

Likely it would be better if the arming prevention reason was set by the code as it prevents arming.  then getArmingPreventionReason() could just return that value, then finally the blink code to look up the mask for the reason in an array or map.

This would also then allow any attached UI (configurator or OSD) to display meaningful status information.

@gaelj Any chance you can do another PR to fix these issues?

**Comment by gaelj** [comment]
No problem @hydra.

**Comment by hydra** [comment]
@gaelj actually, don't.  I'm doing it now as I rebase the OSD stuff onto master.

sorry if you had already started.

**Comment by gaelj** [comment]
I was about to... good think I checked. Go ahead ! :smiley:

**Comment by hydra** [comment]
you can clearly see the dependencies added by all the additional includes.  I missed them in my quick review.

**Comment by hydra** [comment]
the solution involves moving the flight controller specific code into somewhere that has all the dependencies already includes, e.g. in the main mw.c / cleanflight_fc.c file.  likely there's a better place still but that is where the code duplication is too.

**Comment by gaelj** [comment]
Agreed.

**Comment by hydra** [comment]
done, see this first merge (code relocation, removal of dependencies from statusindicator.c)

https://github.com/cleanflight/cleanflight/pull/2226/commits/ecc70776067f7db2aef13efd8b78897908835aa1

then see how it's refactored here (and used by both osd and fc code).

https://github.com/cleanflight/cleanflight/pull/2226/commits/25f4be7377088d38339cee6e479de1d2eb4a3ab4

I added some comments at the top of the file (in lieu of a unit test, *ahem* *cough*).

**Comment by gaelj** [comment]
I suppose those Magic Numbers when you call warningLedCode() will be cleaned up. The first commit is too huge to look at on my phone here.
